app/src/main/java/com/arygm/quickfix/model/category/CategoryRepositoryFirestore.kt

```diff
@@ -91,15 +118,37 @@ open class CategoryRepositoryFirestore(private val db: FirebaseFirestore) : Cate
     return Category(id = id, name = name, description = description, subcategories = subcategories)
   }
 
-  private suspend fun fetchSubcategoriesSuspend(categoryId: String): List<Subcategory>? {
+  private suspend fun fetchSubcategoriesSuspend(categoryId: String): List<Subcategory> {
     return try {
+      // Attempt to fetch subcategories from Room
+      val cachedCategory = dao.getCategoryById(categoryId)
+      if (cachedCategory != null) {
+        val subcategories = cachedCategory.toCategory().subcategories
+        if (subcategories.isNotEmpty()) {
+          Log.d("fetchSubcategoriesSuspend", "Returning cached subcategories for $categoryId")
+          return subcategories
+        }
+      }
+
+      // Fallback to Firestore if no cached data is available
       val subcategoryRef =
           db.collection(collectionPath).document(categoryId).collection("subcategories")
       val result = subcategoryRef.get().await()
-      result.documents.mapNotNull { documentToSubcategory(it) }
+
+      // Map Firestore documents to Subcategory objects
+      val subcategories = result.documents.mapNotNull { documentToSubcategory(it) }
+
+      // Cache fetched subcategories by updating the CategoryEntity in Room
+      cachedCategory?.let {
+        val updatedCategory = it.toCategory().copy(subcategories = subcategories)
+        dao.insertCategory(updatedCategory.toCategoryEntity())
+      }
+
+      Log.d("fetchSubcategoriesSuspend", "Returning Firestore subcategories for $categoryId")
+      subcategories
     } catch (e: Exception) {
-      Log.e("fetchSubcategoriesSuspend", "Error fetching documents from Firestore", e)
-      null
+      Log.e("fetchSubcategoriesSuspend", "Error fetching subcategories for $categoryId", e)
+      emptyList() // Return an empty list in case of an error
     }
```

Comment:
I'm quite confused here, the function does the same thing as before but isn't asynchronous can we just use the one from before but with a correct onSuccess?

Reply:
It is not me that implemented this function but an issue could definitely be opened for this 😉